src/utils/gameStateReducer.ts

```diff
@@ -0,0 +1,84 @@
+import { initialBoard } from 'utils'
+
+const gameStateReducer = (state: GameState, action: Action) => {
+  const { type, newGameState } = action
+
+  switch (type) {
+    case 'RESTART':
+      return {
+        currentBoardState: initialBoard,
+        buttonValue: 0,
+        isWon: false,
+        isTied: false,
+        isXNext: true,
+        winningSquareSet: [],
+        gameHistory: [initialBoard],
+      }
+
+    case 'UPDATE_CURRENT_BOARD':
+      return {
+        ...state,
+        currentBoardState: newGameState.currentBoardState,
+        buttonValue: state.buttonValue + 1,
+      }
+
+    case 'ADD_HISTORY':
+      return {
+        ...state,
+        gameHistory: [
+          ...state.gameHistory.slice(0, state.buttonValue),
+          state.currentBoardState,
+        ],
+      }
+
+    case 'SET_NEXT':
```

Comment:
I would rename this to `NEXT_TURN` and increment the turn here, or just make it generic and allow you to set whatever turn you want. The logic here doesn't really make sense currently, your setting the new value of isXNext based of state that already exists